src/zenml/steps/utils.py

```diff
@@ -410,19 +410,26 @@ def Do(
         return_type: Type[Any] = spec.annotations.get("return", None)
         if return_type is not None:
             if isinstance(return_type, Output):
-                # Resolve named (and multi-) outputs.
-                if len(list(return_type.items())) == 1:
-                    return_values = [return_values]
-                for i, output_tuple in enumerate(return_type.items()):
-                    self.resolve_output_artifact(
-                        output_tuple[0],
-                        output_dict[output_tuple[0]][0],
-                        return_values[i],  # order preserved.
-                    )
+                output_annotations = list(return_type.items())
             else:
-                # Resolve single output
+                output_annotations = [(SINGLE_RETURN_OUT_NAME, return_type)]
+
+            # if there is only one output annotation (either directly specified
+            # or contained in an `Output` tuple) we treat the step function
+            # return value as the return for that output
+            if len(output_annotations) == 1:
+                return_values = [return_values]
+
+            for return_value, (output_name, output_type) in zip(
+                return_values, output_annotations
+            ):
+                if not isinstance(return_value, output_type):
```

Comment:
I was wondering whether we should do this at the materializer base class? The reason is that we might want to extend this in the future to also allow sub-classes etc

Reply:
I thought about that too, but the materializer doesn't actually have access to the type annotation and can in some cases handle multiple unrelated objects (a `List` and an `int` for example). I also think `isinstance` already handles subclasses sufficiently: As long as the returned object is of the specified class or any subclass, it will pass.

Reply:
Ok!